rotkehlchen/chain/ethereum/oracles/uniswap.py

```diff
@@ -0,0 +1,385 @@
+)
+from rotkehlchen.constants.misc import ONE, ZERO
+from rotkehlchen.errors import PriceQueryUnsupportedAsset
+from rotkehlchen.fval import FVal
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEthAddress, Price, Timestamp
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.ethereum.manager import EthereumManager
+
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class UniswapOracle(metaclass=abc.ABCMeta):
+    """
+    Provides shared logic between Uniswap V2 and Uniswap V3 to use them as price oracles.
+    """
+    def __init__(self, eth_manager: 'EthereumManager'):
+        self.eth_manager = eth_manager
+        self.routing_assets = [
+            A_WETH,
+            A_DAI,
+            A_USDT,
+        ]
+
+    @abc.abstractmethod
+    def get_pool(
+        self,
+        token_0: EthereumToken,
+        token_1: EthereumToken,
+    ) -> List[str]:
+        """Given two tokens returns a list of pool where they can be swapped"""
+        raise NotImplementedError('should only be implemented by subclasses')
+
+    @abc.abstractmethod
+    def get_pool_price(
+        self,
+        pool_addr: ChecksumEthAddress,
+        block_identifier: BlockIdentifier = 'latest',
+    ) -> Tuple[FVal, EthereumToken, EthereumToken]:
+        """Retruns the price for the tokens in the given pool and the token0 and
+        token1 of the pool
+        """
+        raise NotImplementedError('should only be implemented by subclasses')
+
+    def rate_limited_in_last(  # pylint: disable=no-self-use
+            self,
+            seconds: Optional[int] = None,  # pylint: disable=unused-argument
+    ) -> bool:
+        return False  # noop for uniswap
+
+    def can_query_history(  # pylint: disable=no-self-use
+        self,
+        from_asset: Asset,  # pylint: disable=unused-argument
+        to_asset: Asset,  # pylint: disable=unused-argument
+        timestamp: Timestamp,  # pylint: disable=unused-argument
+        seconds: Optional[int] = None,  # pylint: disable=unused-argument
+    ) -> bool:
+        return False  # noop for uniswap oracles atm
+
+    def find_route(self, from_asset: EthereumToken, to_asset: EthereumToken) -> List[str]:
+        """
+        Returns a list of the pools that will be needed to go from_asset to to_asset
+        """
+        output = []
+        # If any of the assets is in the glue assets let's see if we find any path
+        # (avoids iterating the list of glue assets)
+        if to_asset in self.routing_assets or from_asset in self.routing_assets:
+            pools = self.get_pool(from_asset, to_asset)
+            for pool in pools:
+                if pool != ZERO_ADDRESS:
+                    return [pool]
+
+        if from_asset == to_asset:
+            return []
+
+        # Try to find one asset that can be used between from_asset and to_asset
+        link_asset = None
```

Comment:
Has it changed since then? I ... ehm .. don't remember how the code looked like more than a month ago :sweat_smile:

Reply:
I've seen how recursion could be used here but it will get messy for what it is and recursion is not exactly what you wanted here. I've added comments and removed duplicated code to make it clearer

Reply:
still not totally convinced. This code seems a bit messy but let's see